---
title: "ratte Guide"
author: "Jianhong Ou"
bibliography: bibliography.bib
csl: nature.csl
vignette: >
  %\VignetteIndexEntry{ratte Vignette}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
    toc_depth: 4
runtime: shiny
---

```{r, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
suppressPackageStartupMessages({
  library(ratte)
  library(rtracklayer)
  library(edgeR)
  library(AnnotationHub)
  library(BSgenome.Drerio.UCSC.danRer11)
  library(TxDb.Drerio.UCSC.danRer11.refGene)
  library(EnhancedVolcano)
  library(pheatmap)
})
knitr::opts_chunk$set(warning=FALSE, message=TRUE, fig.width=6, fig.height=5)
```

# Introduction

Transposable elements (TEs) play a role in fostering genetic diversity,
instigating genome rearrangements, and exerting both positive and negative
effects on the host organism. The regulation of these elements constitutes a
dynamic and intricate process, involving host defense mechanisms to control
their activity.
The [REdiscoverTE](https://github.com/ucsffrancislab/REdiscoverTE), presented in
2019 by Kong et al.[@kong2019transposable], showcased remarkable accuracy in TE
quantification. It stands out as a significantly more comprehensive and accurate
method for robust, whole-transcriptome quantification.

Nevertheless, the existing _REdiscoverTE_ is constrained by the human genome
assembly GRCh38. To broaden the scope of annotation to encompass more species,
the _ratte_ package has been devised. This package leverages the resources
available within the Bioconductor community to prepare the necessary sequence
files and annotations.

# Quick start for RNA-seq data

Here is an example using _ratte_ with a subset of RNA-seq data.

## Installation

First, install _ratte_ and other packages required to run 
the examples.
Please note that the example dataset used here is from zebrafish.
To run analysis with dataset from a different species or different assembly,
please install the corresponding Bsgenome and TxDb.
For example, to analyze mouse data aligned to mm10, please install `BSgenome.Mmusculus.UCSC.mm10`, and `TxDb.Mmusculus.UCSC.mm10.knownGene`.
You can also generate a TxDb object by functions `makeTxDbFromGFF` from
a local gff file, or `makeTxDbFromUCSC`, `makeTxDbFromBiomart`, and
`makeTxDbFromEnsembl`, from online resources in _GenomicFeatures_ package.
The BSgenome package can be created by _BSgenomeForge_ package.

Please note that the example dataset used here is from a small subset of
zebrafish fin regeneration[@thompson2020identification].

```{r, eval=FALSE}
library(devtools)
devtools::install("jianhong/ratte")

## other annotation and helper libraries
library(BiocManager)
BiocManager::install(c("AnnotationHub", "rtracklayer",
                       "BSgenome.Drerio.UCSC.danRer11",
                       "TxDb.Drerio.UCSC.danRer11.refGene",
                       "EnhancedVolcano",
                       "pheatmap"))
```

## Load library

```{r}
library(ratte)
```

## Generating reference transcriptome

The reference transcriptome includes:

1. Distinct RNA transcript sequences prepared by `GenomicFeatures::extractTranscriptSeqs` function for output of `GenomicFeatures::exonsBy` at transcript level.

2. Distinct sequences representing RE-containing introns by `GenomicFeatures::intronicParts`.

3. Distinct RepeatMasker elements in the intergenic region.

In this manual, we will use the fish genome.
```{r}
library(AnnotationHub)
library(BSgenome.Drerio.UCSC.danRer11)
## set genome, Drerio is a shortname for BSgenome.Drerio.UCSC.danRer11
genome <- Drerio
library(TxDb.Drerio.UCSC.danRer11.refGene)
txdb <- TxDb.Drerio.UCSC.danRer11.refGene ## give it a short name
```

```{r}
ah <- AnnotationHub()
(rmsk <- query(ah, c('RepeatMasker', 'Zebrafish')))
```
```{r, eval=FALSE}
rmsk <- rmsk[['AH98980']]
```

We will use a subset of fish genome in this manual.
Please skip this step if you want try the full fish genome.
```{r}
rmsk <- readRDS(system.file('extdata', 'danRer11_rmsk_chr25_sub.RDS',
                            package='ratte'))
```

The `prepareSeq` function will extract the sequences and merge the 
RepeatMasker elements annotation.
The `saveFasta` function is uesed to export the sequences into a fasta file.

```{r}
seq_anno <- prepareSeq(genome, txdb, rmsk,
## The subsetGRanges parameter is only used for subset the genome.
## Please remove this parameter for full fish genome.
                       subsetGRanges=GRanges('chr25:11000000-15000000'))
rna_outfile <- tempfile(fileext='.fa.gz')
saveFasta(seq_anno, rna_outfile)
```

## Align reads by salmon

First we will create salmon index. In this example, we will just export the 
command by setting `dryrun` parameter to TRUE.
```{r}
prepareSalmonIndex(salmonPath='salmon',
                   FaFile=rna_outfile,
                   index='RNA_TEindex',
                   dryrun=TRUE)
```

Second, we will align reads to salmon index.
```{r}
alignReadsBySalmon('sample1.R1.fq.gz', 'sample1.R2.fq.gz',
                    output='sample1_salmon_output',
                    index='RNA_TEindex',
                    dryrun=TRUE)
```

Once alignment is done, we will get quantification file named as `quant.sf`.
In this manual, we will load pre-aligned files.
```{r}
quant_sfs <- dir(system.file('extdata', 'RNA-seq', package='ratte'), 'quant.sf',
                 full.names=TRUE)
qnames <- sub("^.*\\/(.*?).quant.sf", "\\1", quant_sfs)
```

## Aggregate alignments to repeat element (RE).

The `createDGELists` function will extract counts number from the `quant.sf`
files and generate a list of `edgeR::DGEList` object. The available choices for
`select_feature` are 'all', 'exon', 'intron', 'gene', and 'intergenic'.
Here we use `gene` as an example.
Please note that the DGEList library size and normliazation factor will be
determined by gene level counts.

```{r}
ys <- createDGELists(quant_sfs, qnames, seq_anno,
                     select_feature = 'gene',
                     group=sub('.rep.', '', qnames),
                     remove.zeros = TRUE)
names(ys)
ys[[1]]
```

## Analyze results with _edgeR_


```{r}
library(edgeR)
library(EnhancedVolcano)
library(pheatmap)
y <- ys[['repName']]
keep <- filterByExpr(y)
y <- y[keep, keep.lib.sizes=TRUE]
design <- model.matrix(~y$samples$group)
y <- estimateDisp(y, design)
fit <- glmFit(y, design)
lrt <- glmLRT(fit, coef=2)
res <- as.data.frame(topTags(lrt, n=nrow(lrt)))
EnhancedVolcano(res,
                lab=rownames(res),
                x='logFC',
                y='PValue',
                title='Volcano plot for repeat elements',
                subtitle = NULL,
                legendPosition = 'right'
                )

de <- res[abs(res$logFC)>2 & res$FDR<0.05, ]
plotdata <- y$tpm[rownames(de), ]
pheatmap(plotdata, scale = 'row', cluster_cols = FALSE)
```

# Quick start for ATAC-seq/ChIP-seq data

## Generating reference for ATAC-seq/ChIP-seq

The reference are distinct sequences representing RE-containing peaks.

For ATAC-seq/ChIP-seq, to prepare the sequences, the called peaks are required.
```{r}
library(rtracklayer)
peaks <- import(system.file('extdata', 'merged.peaks.bed.gz',
                            package = 'ratte'),
                format = 'bed')
atac_seq_anno <- prepareSeq(genome, txdb, rmsk, peaks)
atac_outfile <- tempfile(fileext='.fa.gz')
saveFasta(atac_seq_anno, atac_outfile)
```

## Align reads by salmon

First we will create salmon index. In this example, we will just export the 
command by setting `dryrun` parameter to TRUE.
```{r}
prepareSalmonIndex(salmonPath='salmon',
                   FaFile=atac_outfile,
                   index='ATAC_TEindex',
                   dryrun=TRUE)
```

Second, we will align reads to salmon index.
```{r}
alignReadsBySalmon('sample1.R1.fq.gz', 'sample1.R2.fq.gz',
                    output='sample1_salmon_output',
                    index='ATAC_TEindex',
                    dryrun=TRUE)
```

Once alignment is done, we will get quantification file named as `quant.sf`.
In this manual, we will load pre-aligned files.
```{r}
atac_quant_sfs <- dir(system.file('extdata', 'ATAC-seq',
                                  package='ratte'),
                      'quant.sf',
                      full.names=TRUE)
atac_qnames <- sub("^.*\\/(.*?).quant.sf", "\\1", atac_quant_sfs)
```

## Aggregate alignments to repeat element (RE).

The `createDGELists` function will extract counts number from the `quant.sf`
files and generate a list of `edgeR::DGEList` object. The available choices for
`select_feature` are 'all', 'exon', 'intron', 'gene', and 'intergenic'.
Here we use `all` as an example.

```{r}
atac_ys <- createDGELists(atac_quant_sfs, atac_qnames, atac_seq_anno,
                          select_feature = 'peak',
                          group=sub('.rep.', '', atac_qnames),
                          norm_level='peak',
                          remove.zeros = TRUE)
names(atac_ys)
atac_ys[[1]]
```

## Analyze results with _edgeR_

```{r}
y <- atac_ys[['repName']]
keep <- filterByExpr(y)
y <- y[keep, keep.lib.sizes=TRUE]
design <- model.matrix(~y$samples$group)
y <- estimateDisp(y, design)
fit <- glmFit(y, design)
lrt <- glmLRT(fit, coef=2)
atac_res <- as.data.frame(topTags(lrt, n=nrow(lrt)))
EnhancedVolcano(atac_res,
                lab=rownames(atac_res),
                x='logFC',
                y='PValue',
                title='Volcano plot for repeat elements',
                subtitle = NULL,
                legendPosition = 'right'
                )
## this is very small test data, we set the cutoff a small number.
atac_de <- atac_res[abs(atac_res$logFC)>log2(1.5) & atac_res$PValue<0.05, ]
plotdata <- y$tpm[rownames(atac_de), ]
pheatmap(plotdata, scale = 'row', cluster_cols = FALSE)
```

# Merge the results from RNA-seq and ATAC-seq

## For repeat elements level
```{r}
m <- merge(res, atac_res, suffixes = c(".RNA",".ATAC"),
           by='row.names')
m <- m[order(-1*sign(m$logFC.RNA)*sign(m$logFC.ATAC), 
               sqrt(m$PValue.RNA^2 + m$PValue.ATAC^2)), ]
head(m, n=2)
head(seq_anno[seq_anno$repName==m[1, 1], ])
head(atac_seq_anno[atac_seq_anno$repName==m[1, 1], ])
```

## For transcripts level

```{r}
DE <- function(y){
  keep <- filterByExpr(y)
  y <- y[keep, keep.lib.sizes=TRUE]
  design <- model.matrix(~y$samples$group)
  y <- estimateDisp(y, design)
  fit <- glmFit(y, design)
  lrt <- glmLRT(fit, coef=2)
  res <- as.data.frame(topTags(lrt, n=nrow(lrt)))
  ## for ATAC-seq/ChIP-seq, the tx_name were annotated by the nearest
  ## transcript, you may want to change this with other annotation.
  tx_name <- strsplit(res$tx_name, split=',')
  tx_name <- lapply(tx_name, unique)
  res <- res[rep(seq.int(nrow(res)), lengths(tx_name)), ]
  res$tx_name <- unlist(tx_name)
  res
}
tx.RNA <- DE(ys[['rawCounts']])
tx.ATAC <- DE(atac_ys[['rawCounts']])
head(tx.RNA[, colnames(tx.RNA)!='seq'], n=2)
head(tx.ATAC[, colnames(tx.ATAC)!='seq'], n=2)
tx_level <- merge(tx.RNA, tx.ATAC,
                  by=c('repName', 'repClass', 'repFamily', 'tx_name'),
                  suffix=c('.RNA', '.ATAC'))
tx_level[1:2, c('tx_name', 'repName', 'feature.RNA',
                'logFC.RNA', 'PValue.RNA', 'FDR.RNA',
                'logFC.ATAC', 'PValue.ATAC', 'FDR.ATAC')]
```

# SessionInfo

```{r}
sessionInfo()
```


# References
