<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ratte Guide • ratte</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ratte Guide">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ratte</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/ratte.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jianhong/ratte" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>ratte Guide</h1>
                        <h4 data-toc-skip class="author">Jianhong
Ou</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jianhong/ratte/blob/main/vignettes/ratte.Rmd" class="external-link"><code>vignettes/ratte.Rmd</code></a></small>
      <div class="hidden name"><code>ratte.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Transposable elements (TEs) play a role in fostering genetic
diversity, instigating genome rearrangements, and exerting both positive
and negative effects on the host organism. The regulation of these
elements constitutes a dynamic and intricate process, involving host
defense mechanisms to control their activity. The <a href="https://github.com/ucsffrancislab/REdiscoverTE" class="external-link">REdiscoverTE</a>,
presented in 2019 by Kong et al.<span class="citation"><sup>1</sup></span>, showcased remarkable accuracy in
TE quantification. It stands out as a significantly more comprehensive
and accurate method for robust, whole-transcriptome quantification.</p>
<p>Nevertheless, the existing <em>REdiscoverTE</em> is constrained by
the human genome assembly GRCh38. To broaden the scope of annotation to
encompass more species, the <em>ratte</em> package has been devised.
This package leverages the resources available within the Bioconductor
community to prepare the necessary sequence files and annotations.</p>
</div>
<div class="section level2">
<h2 id="quick-start-for-rna-seq-or-scrna-seq-data">Quick start for RNA-seq or scRNA-seq data<a class="anchor" aria-label="anchor" href="#quick-start-for-rna-seq-or-scrna-seq-data"></a>
</h2>
<p>Here is an example using <em>ratte</em> with a subset of RNA-seq
data.</p>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>First, install <em>ratte</em> and other packages required to run the
examples. Please note that the example dataset used here is from
zebrafish. To run analysis with dataset from a different species or
different assembly, please install the corresponding BSgenome and TxDb.
For example, to analyze mouse data aligned to mm10, please install
<code>BSgenome.Mmusculus.UCSC.mm10</code>, and
<code>TxDb.Mmusculus.UCSC.mm10.knownGene</code>. You can also generate a
TxDb object by functions <code>makeTxDbFromGFF</code> from a local gff
file, or <code>makeTxDbFromUCSC</code>,
<code>makeTxDbFromBiomart</code>, and <code>makeTxDbFromEnsembl</code>,
from online resources in <em>GenomicFeatures</em> package. The BSgenome
package can be created by <em>BSgenomeForge</em> package.</p>
<p>Please note that the example dataset used here is from a small subset
of zebrafish fin regeneration<span class="citation"><sup>2</sup></span>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"jianhong/ratte"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## other annotation and helper libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">BiocManager</a></span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AnnotationHub"</span>, <span class="st">"rtracklayer"</span>,</span>
<span>                       <span class="st">"BSgenome.Drerio.UCSC.danRer11"</span>,</span>
<span>                       <span class="st">"TxDb.Drerio.UCSC.danRer11.refGene"</span>,</span>
<span>                       <span class="st">"EnhancedVolcano"</span>,</span>
<span>                       <span class="st">"pheatmap"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-library">Load library<a class="anchor" aria-label="anchor" href="#load-library"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jianhong/ratte" class="external-link">ratte</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="generating-reference-transcriptome">Generating reference transcriptome<a class="anchor" aria-label="anchor" href="#generating-reference-transcriptome"></a>
</h3>
<p>The reference transcriptome includes:</p>
<ol style="list-style-type: decimal">
<li><p>Distinct RNA transcript sequences prepared by
<code><a href="https://rdrr.io/pkg/GenomicFeatures/man/extractTranscriptSeqs.html" class="external-link">GenomicFeatures::extractTranscriptSeqs</a></code> function for output
of <code><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html" class="external-link">GenomicFeatures::exonsBy</a></code> at transcript level.</p></li>
<li><p>Distinct sequences representing RE-containing introns by
<code><a href="https://rdrr.io/pkg/GenomicFeatures/man/exonicParts.html" class="external-link">GenomicFeatures::intronicParts</a></code>.</p></li>
<li><p>Distinct RepeatMasker elements in the intergenic region.</p></li>
</ol>
<p>In this manual, we will use the fish genome.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Drerio.UCSC.danRer11</span><span class="op">)</span></span>
<span><span class="co">## set genome, Drerio is a shortname for BSgenome.Drerio.UCSC.danRer11</span></span>
<span><span class="va">genome</span> <span class="op">&lt;-</span> <span class="va">Drerio</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Drerio.UCSC.danRer11.refGene</span><span class="op">)</span></span>
<span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Drerio.UCSC.danRer11.refGene</span> <span class="co">## give it a short name</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ah</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">rmsk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">ah</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'RepeatMasker'</span>, <span class="st">'Zebrafish'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## AnnotationHub with 1 record</span></span>
<span><span class="co">## # snapshotDate(): 2025-06-23</span></span>
<span><span class="co">## # names(): AH98980</span></span>
<span><span class="co">## # $dataprovider: UCSC</span></span>
<span><span class="co">## # $species: Danio rerio</span></span>
<span><span class="co">## # $rdataclass: GRanges</span></span>
<span><span class="co">## # $rdatadateadded: 2022-01-31</span></span>
<span><span class="co">## # $title: UCSC RepeatMasker annotations (Nov2017) for Zebrafish (danRer11)</span></span>
<span><span class="co">## # $description: UCSC RepeatMasker annotations (Nov2017) for Zebrafish -Danio...</span></span>
<span><span class="co">## # $taxonomyid: 7955</span></span>
<span><span class="co">## # $genome: danRer11</span></span>
<span><span class="co">## # $sourcetype: UCSC track</span></span>
<span><span class="co">## # $sourceurl: https://functionalgenomics.upf.edu/annotationhub/repeatmasker/...</span></span>
<span><span class="co">## # $sourcesize: NA</span></span>
<span><span class="co">## # $tags: c("RepeatMasker", "UCSC") </span></span>
<span><span class="co">## # retrieve record with 'object[["AH98980"]]'</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rmsk</span> <span class="op">&lt;-</span> <span class="va">rmsk</span><span class="op">[[</span><span class="st">'AH98980'</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>We will use a subset of fish genome in this manual. Please skip this
step if you want try the full fish genome.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rmsk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'danRer11_rmsk_chr25_sub.RDS'</span>,</span>
<span>                            package<span class="op">=</span><span class="st">'ratte'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The <code>prepareSeq</code> function will extract the sequences and
merge the RepeatMasker elements annotation. The <code>saveFasta</code>
function is uesed to export the sequences into a fasta file.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareSeq.html">prepareSeq</a></span><span class="op">(</span><span class="va">genome</span>, <span class="va">txdb</span>, <span class="va">rmsk</span>,</span>
<span><span class="co">## The subsetGRanges parameter is only used for subset the genome.</span></span>
<span><span class="co">## Please remove this parameter for full fish genome.</span></span>
<span>                       subsetGRanges<span class="op">=</span><span class="fu">GRanges</span><span class="op">(</span><span class="st">'chr25:11000000-15000000'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Please note that the seq_anno will be used in later. </span></span>
<span><span class="co">## It will be good to save it as an R object by</span></span>
<span><span class="co">## saveRDS(seq_anno, 'seq_anno.rds')</span></span>
<span><span class="va">rna_outfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext<span class="op">=</span><span class="st">'.fa.gz'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/saveFasta.html">saveFasta</a></span><span class="op">(</span><span class="va">seq_anno</span>, <span class="va">rna_outfile</span><span class="op">)</span></span></code></pre></div>
<p>For single cell RNA-seq data, a transcript to gene map file is
required.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq2name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>s<span class="op">=</span><span class="va">seq_anno</span><span class="op">$</span><span class="va">seqname</span>, n<span class="op">=</span><span class="va">seq_anno</span><span class="op">$</span><span class="va">repName</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">seq2name</span>, <span class="st">'seq2name.tsv'</span>,</span>
<span>            row.names <span class="op">=</span> <span class="cn">FALSE</span>, col.names <span class="op">=</span> <span class="cn">FALSE</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            sep <span class="op">=</span> <span class="st">'\t'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="align-reads-by-salmon">Align reads by salmon<a class="anchor" aria-label="anchor" href="#align-reads-by-salmon"></a>
</h3>
<p>First we will create salmon index. In this example, we will just
export the command by setting <code>dryrun</code> parameter to TRUE.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prepareSalmonIndex.html">prepareSalmonIndex</a></span><span class="op">(</span>salmonPath<span class="op">=</span><span class="st">'salmon'</span>,</span>
<span>                   FaFile<span class="op">=</span><span class="va">rna_outfile</span>,</span>
<span>                   index<span class="op">=</span><span class="st">'RNA_TEindex'</span>,</span>
<span>                   dryrun<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## salmon index --threads 2 -t /tmp/RtmpecBqKk/file1f6a36b9f631.fa.gz -i RNA_TEindex -n</span></span></code></pre>
<p>Second, we will align reads to salmon index.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/alignReadsBySalmon.html">alignReadsBySalmon</a></span><span class="op">(</span><span class="st">'sample1.R1.fq.gz'</span>, <span class="st">'sample1.R2.fq.gz'</span>,</span>
<span>                    output<span class="op">=</span><span class="st">'sample1_salmon_output'</span>,</span>
<span>                    index<span class="op">=</span><span class="st">'RNA_TEindex'</span>,</span>
<span>                    dryrun<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## salmon quant --seqBias --gcBias --index RNA_TEindex --libType A --validateMappings --threads 2 -o sample1_salmon_output -1 sample1.R1.fq.gz -2 sample1.R2.fq.gz</span></span></code></pre>
<p>For single cell data, there are multiple additional parameters are
required.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/alignReadsBySalmon.html">alignReadsBySalmon</a></span><span class="op">(</span><span class="st">'sample1.R1.fq.gz'</span>, <span class="st">'sample1.R2.fq.gz'</span>,</span>
<span>                   output<span class="op">=</span><span class="st">'sample1_salmon_output'</span>,</span>
<span>                   index<span class="op">=</span><span class="st">'RNA_TEindex'</span>,</span>
<span>                   singlecell_protocol <span class="op">=</span> <span class="st">'chromiumV3'</span>,<span class="co">## if your 10x kit ver &gt;= 3 </span></span>
<span>                   tgMap<span class="op">=</span><span class="st">'seq2name.tsv'</span>, <span class="co">## saved by previous step</span></span>
<span>                   whitelist<span class="op">=</span><span class="st">'cellRanger/outs/filtered_feature_bc_matrix/barcodes.tsv.gz'</span>,</span>
<span>                   dryrun<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Once alignment is done, we will get quantification file named as
<code>quant.sf</code>. In this manual, we will load pre-aligned
files.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">quant_sfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'RNA-seq'</span>, package<span class="op">=</span><span class="st">'ratte'</span><span class="op">)</span>, <span class="st">'quant.sf'</span>,</span>
<span>                 full.names<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">qnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"^.*\\/(.*?).quant.sf"</span>, <span class="st">"\\1"</span>, <span class="va">quant_sfs</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="aggregate-alignments-to-repeat-element-re-">Aggregate alignments to repeat element (RE).<a class="anchor" aria-label="anchor" href="#aggregate-alignments-to-repeat-element-re-"></a>
</h3>
<p>The <code>createDGELists</code> function will extract counts number
from the <code>quant.sf</code> files and generate a list of
<code><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">edgeR::DGEList</a></code> object. The available choices for
<code>select_feature</code> are ‘all’, ‘exon’, ‘intron’, ‘gene’, and
‘intergenic’. Here we use <code>gene</code> as an example. Please note
that the DGEList library size and normliazation factor will be
determined by gene level counts.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ys</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createDGELists.html">createDGELists</a></span><span class="op">(</span><span class="va">quant_sfs</span>, <span class="va">qnames</span>, <span class="va">seq_anno</span>,</span>
<span>                     select_feature <span class="op">=</span> <span class="st">'gene'</span>,</span>
<span>                     group<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">'.rep.'</span>, <span class="st">''</span>, <span class="va">qnames</span><span class="op">)</span>,</span>
<span>                     remove.zeros <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Removing 26 rows with all zero counts</span></span></code></pre>
<pre><code><span><span class="co">## Removing 1 rows with all zero counts</span></span></code></pre>
<pre><code><span><span class="co">## Removing 4 rows with all zero counts</span></span></code></pre>
<pre><code><span><span class="co">## Removing 60 rows with all zero counts</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ys</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "repName"   "repClass"  "repFamily" "rawCounts"</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ys</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "DGEList"</span></span>
<span><span class="co">## $counts</span></span>
<span><span class="co">##           Reg.rep1 Reg.rep2 Uni.rep1 Uni.rep2</span></span>
<span><span class="co">## (A)n        22.143   27.214   27.546   18.000</span></span>
<span><span class="co">## (AACTGA)n   11.239   17.000   11.000    4.000</span></span>
<span><span class="co">## (AAT)n      11.000    7.000   47.000   46.000</span></span>
<span><span class="co">## (AATA)n      6.006    6.000   11.529    4.001</span></span>
<span><span class="co">## (AATAA)n     3.000    1.000    4.000    3.000</span></span>
<span><span class="co">## 350 more rows ...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $samples</span></span>
<span><span class="co">##          group  lib.size norm.factors</span></span>
<span><span class="co">## Reg.rep1   Reg  5778.507            1</span></span>
<span><span class="co">## Reg.rep2   Reg  6219.626            1</span></span>
<span><span class="co">## Uni.rep1   Uni 11454.850            1</span></span>
<span><span class="co">## Uni.rep2   Uni 10224.091            1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $tpm</span></span>
<span><span class="co">##            Reg.rep1  Reg.rep2  Uni.rep1  Uni.rep2</span></span>
<span><span class="co">## (A)n      18.625446 19.844470 28.515103 19.587027</span></span>
<span><span class="co">## (AACTGA)n  5.391495  6.781941  5.878126  2.179567</span></span>
<span><span class="co">## (AAT)n     9.644679 11.832648 63.671101 68.940900</span></span>
<span><span class="co">## (AATA)n    8.228509  9.468091 18.180161  7.120650</span></span>
<span><span class="co">## (AATAA)n   3.411824  0.443921  3.333512  3.882252</span></span>
<span><span class="co">## 350 more rows ...</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="analyze-results-with-edger">Analyze results with <em>edgeR</em><a class="anchor" aria-label="anchor" href="#analyze-results-with-edger"></a>
</h3>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kevinblighe/EnhancedVolcano" class="external-link">EnhancedVolcano</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">ys</span><span class="op">[[</span><span class="st">'repName'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">keep</span>, keep.lib.sizes<span class="op">=</span><span class="cn">TRUE</span><span class="op">]</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html" class="external-link">glmFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">lrt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmLRT.html" class="external-link">glmLRT</a></span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">lrt</span>, n<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">lrt</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EnhancedVolcano/man/EnhancedVolcano.html" class="external-link">EnhancedVolcano</a></span><span class="op">(</span><span class="va">res</span>,</span>
<span>                lab<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>,</span>
<span>                x<span class="op">=</span><span class="st">'logFC'</span>,</span>
<span>                y<span class="op">=</span><span class="st">'PValue'</span>,</span>
<span>                title<span class="op">=</span><span class="st">'Volcano plot for repeat elements'</span>,</span>
<span>                subtitle <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                legendPosition <span class="op">=</span> <span class="st">'right'</span></span>
<span>                <span class="op">)</span></span></code></pre></div>
<p><img src="ratte_files/figure-html/unnamed-chunk-15-1.png" width="576"></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">2</span> <span class="op">&amp;</span> <span class="va">res</span><span class="op">$</span><span class="va">FDR</span><span class="op">&lt;</span><span class="fl">0.05</span>, <span class="op">]</span></span>
<span><span class="va">plotdata</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">$</span><span class="va">tpm</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">de</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">plotdata</span>, scale <span class="op">=</span> <span class="st">'row'</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="ratte_files/figure-html/unnamed-chunk-15-2.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="aggregate-counts-for-scrna-seq">Aggregate counts for scRNA-seq<a class="anchor" aria-label="anchor" href="#aggregate-counts-for-scrna-seq"></a>
</h3>
<p>We will use <code>tximeta</code> to import the counts number for the
alevin quants. And downstream will treat it same as single cell RNA-seq
data with the tools such as Seurat or scanpy.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tximeta</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">quants_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="st">'alevin_output'</span>, <span class="st">'quants_mat.gz'</span>,</span>
<span>                    recursive <span class="op">=</span> <span class="cn">TRUE</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">quants_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'WT1'</span>, <span class="st">'WT2'</span>, <span class="st">'KD3'</span>, <span class="st">'KD4'</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">quants_files</span>, <span class="va">tximeta</span>, type<span class="op">=</span><span class="st">'alevin'</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="va">se</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>, FUN<span class="op">=</span><span class="kw">function</span><span class="op">(</span><span class="va">.ele</span>, <span class="va">.name</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">CreateSeuratObject</span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">.ele</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span>,</span>
<span>                     project <span class="op">=</span> <span class="va">.name</span>,</span>
<span>                     min.cells <span class="op">=</span> <span class="fl">0</span>, min.features <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">## simple merge</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">seu</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">seu</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## filter cells</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">VlnPlot</span><span class="op">(</span><span class="va">seu</span>, features<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"QC.pdf"</span>, plot<span class="op">=</span><span class="va">p</span>, width<span class="op">=</span><span class="fl">7</span>, height<span class="op">=</span><span class="fl">7</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">seu</span>, subset<span class="op">=</span><span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">200</span> <span class="op">&amp;</span> <span class="va">nCount_RNA</span><span class="op">&lt;</span><span class="fl">30000</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">VlnPlot</span><span class="op">(</span><span class="va">seu</span>, features<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"QC2.pdf"</span>, plot<span class="op">=</span><span class="va">p</span>, width<span class="op">=</span><span class="fl">7</span>, height<span class="op">=</span><span class="fl">7</span><span class="op">)</span></span>
<span><span class="co">## standard pipeline</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">seu</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="va">seu</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">3000</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">ScaleData</span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">seu</span>, features <span class="op">=</span> <span class="fu">VariableFeatures</span><span class="op">(</span>object <span class="op">=</span> <span class="va">seu</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">seu</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">'WT'</span>, <span class="va">seu</span><span class="op">$</span><span class="va">orig.ident</span><span class="op">)</span>, <span class="st">'WT'</span>, <span class="st">'KD'</span><span class="op">)</span></span>
<span><span class="fu">Idents</span><span class="op">(</span><span class="va">seu</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"group"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">Idents</span><span class="op">(</span><span class="va">seu</span><span class="op">)</span><span class="op">)</span><span class="co">## confirm the idents.</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">JoinLayers</span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu">FindMarkers</span><span class="op">(</span><span class="va">seu</span>, ident.1<span class="op">=</span><span class="st">'WT'</span>, ident.2<span class="op">=</span><span class="st">'KD'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">markers</span>, <span class="st">'TE.DE.csv'</span><span class="op">)</span></span>
<span><span class="co">## psuedobulk</span></span>
<span><span class="va">pseudo_</span> <span class="op">&lt;-</span> <span class="fu">AggregateExpression</span><span class="op">(</span><span class="va">seu</span>, assays<span class="op">=</span><span class="st">"RNA"</span>,</span>
<span>                               return.seurat<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                               group.by<span class="op">=</span><span class="st">'orig.ident'</span><span class="op">)</span></span>
<span><span class="va">pseudo_</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">'WT'</span>, <span class="va">pseudo_</span><span class="op">$</span><span class="va">orig.ident</span><span class="op">)</span>, <span class="st">'WT'</span>, <span class="st">'KD'</span><span class="op">)</span></span>
<span><span class="fu">Idents</span><span class="op">(</span><span class="va">pseudo_</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'group'</span></span>
<span><span class="va">bulk_markers</span> <span class="op">&lt;-</span> <span class="fu">FindMarkers</span><span class="op">(</span><span class="va">pseudo_</span>, ident.1 <span class="op">=</span> <span class="st">'WT'</span>, ident.2 <span class="op">=</span> <span class="st">'KD'</span>,</span>
<span>                            test.use<span class="op">=</span><span class="st">"DESeq2"</span>, min.cells.group <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="quick-start-for-atac-seqchip-seq-data">Quick start for ATAC-seq/ChIP-seq data<a class="anchor" aria-label="anchor" href="#quick-start-for-atac-seqchip-seq-data"></a>
</h2>
<div class="section level3">
<h3 id="generating-reference-for-atac-seqchip-seq">Generating reference for ATAC-seq/ChIP-seq<a class="anchor" aria-label="anchor" href="#generating-reference-for-atac-seqchip-seq"></a>
</h3>
<p>The reference are distinct sequences representing RE-containing
peaks.</p>
<p>For ATAC-seq/ChIP-seq, to prepare the sequences, the called peaks are
required.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'merged.peaks.bed.gz'</span>,</span>
<span>                            package <span class="op">=</span> <span class="st">'ratte'</span><span class="op">)</span>,</span>
<span>                format <span class="op">=</span> <span class="st">'bed'</span><span class="op">)</span></span>
<span><span class="va">atac_seq_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareSeq.html">prepareSeq</a></span><span class="op">(</span><span class="va">genome</span>, <span class="va">txdb</span>, <span class="va">rmsk</span>, <span class="va">peaks</span><span class="op">)</span></span>
<span><span class="va">atac_outfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext<span class="op">=</span><span class="st">'.fa.gz'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/saveFasta.html">saveFasta</a></span><span class="op">(</span><span class="va">atac_seq_anno</span>, <span class="va">atac_outfile</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="align-reads-by-salmon-1">Align reads by salmon<a class="anchor" aria-label="anchor" href="#align-reads-by-salmon-1"></a>
</h3>
<p>First we will create salmon index. In this example, we will just
export the command by setting <code>dryrun</code> parameter to TRUE.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prepareSalmonIndex.html">prepareSalmonIndex</a></span><span class="op">(</span>salmonPath<span class="op">=</span><span class="st">'salmon'</span>,</span>
<span>                   FaFile<span class="op">=</span><span class="va">atac_outfile</span>,</span>
<span>                   index<span class="op">=</span><span class="st">'ATAC_TEindex'</span>,</span>
<span>                   dryrun<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## salmon index --threads 2 -t /tmp/RtmpecBqKk/file1f6a7d3b0462.fa.gz -i ATAC_TEindex -n</span></span></code></pre>
<p>Second, we will align reads to salmon index.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/alignReadsBySalmon.html">alignReadsBySalmon</a></span><span class="op">(</span><span class="st">'sample1.R1.fq.gz'</span>, <span class="st">'sample1.R2.fq.gz'</span>,</span>
<span>                    output<span class="op">=</span><span class="st">'sample1_salmon_output'</span>,</span>
<span>                    index<span class="op">=</span><span class="st">'ATAC_TEindex'</span>,</span>
<span>                    dryrun<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## salmon quant --seqBias --gcBias --index ATAC_TEindex --libType A --validateMappings --threads 2 -o sample1_salmon_output -1 sample1.R1.fq.gz -2 sample1.R2.fq.gz</span></span></code></pre>
<p>Once alignment is done, we will get quantification file named as
<code>quant.sf</code>. In this manual, we will load pre-aligned
files.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">atac_quant_sfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'ATAC-seq'</span>,</span>
<span>                                  package<span class="op">=</span><span class="st">'ratte'</span><span class="op">)</span>,</span>
<span>                      <span class="st">'quant.sf'</span>,</span>
<span>                      full.names<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">atac_qnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"^.*\\/(.*?).quant.sf"</span>, <span class="st">"\\1"</span>, <span class="va">atac_quant_sfs</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="aggregate-alignments-to-repeat-element-re--1">Aggregate alignments to repeat element (RE).<a class="anchor" aria-label="anchor" href="#aggregate-alignments-to-repeat-element-re--1"></a>
</h3>
<p>The <code>createDGELists</code> function will extract counts number
from the <code>quant.sf</code> files and generate a list of
<code><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">edgeR::DGEList</a></code> object. The available choices for
<code>select_feature</code> are ‘all’, ‘exon’, ‘intron’, ‘gene’, and
‘intergenic’. Here we use <code>all</code> as an example.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">atac_ys</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createDGELists.html">createDGELists</a></span><span class="op">(</span><span class="va">atac_quant_sfs</span>, <span class="va">atac_qnames</span>, <span class="va">atac_seq_anno</span>,</span>
<span>                          select_feature <span class="op">=</span> <span class="st">'peak'</span>,</span>
<span>                          group<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">'.rep.'</span>, <span class="st">''</span>, <span class="va">atac_qnames</span><span class="op">)</span>,</span>
<span>                          norm_level<span class="op">=</span><span class="st">'peak'</span>,</span>
<span>                          remove.zeros <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Removing 37 rows with all zero counts</span></span></code></pre>
<pre><code><span><span class="co">## Removing 727 rows with all zero counts</span></span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">atac_ys</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "repName"   "repClass"  "repFamily" "rawCounts"</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">atac_ys</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "DGEList"</span></span>
<span><span class="co">## $counts</span></span>
<span><span class="co">##          Reg.rep1 Reg.rep2 Reg.rep3 Uni.rep1 Uni.rep2 Uni.rep3</span></span>
<span><span class="co">## (A)n       99.994      125  121.979  129.369      106   95.878</span></span>
<span><span class="co">## (AAA)n      7.000        7    9.000    9.000        8    6.000</span></span>
<span><span class="co">## (AAAAT)n    4.000        4    2.000    9.000        3    1.000</span></span>
<span><span class="co">## (AAAC)n     0.000        0    4.000    0.000        1    1.000</span></span>
<span><span class="co">## (AAAT)n    12.000       16    9.000   18.000        9   10.000</span></span>
<span><span class="co">## 625 more rows ...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $samples</span></span>
<span><span class="co">##          group lib.size norm.factors</span></span>
<span><span class="co">## Reg.rep1   Reg 12166.45            1</span></span>
<span><span class="co">## Reg.rep2   Reg 12340.34            1</span></span>
<span><span class="co">## Reg.rep3   Reg 12775.16            1</span></span>
<span><span class="co">## Uni.rep1   Uni 13935.17            1</span></span>
<span><span class="co">## Uni.rep2   Uni 11293.21            1</span></span>
<span><span class="co">## Uni.rep3   Uni 11001.10            1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $tpm</span></span>
<span><span class="co">##            Reg.rep1   Reg.rep2   Reg.rep3   Uni.rep1   Uni.rep2    Uni.rep3</span></span>
<span><span class="co">## (A)n     6928.11611 6042.92255 4735.13835 6232.70360 6676.89490 5076.559294</span></span>
<span><span class="co">## (AAA)n     26.39531   24.96572   36.47862   28.37995   32.61850   23.779377</span></span>
<span><span class="co">## (AAAAT)n   26.38979   27.27461   15.11700   59.20299   23.74255    7.120842</span></span>
<span><span class="co">## (AAAC)n     0.00000    0.00000 2375.17580    0.00000  704.50618  680.073838</span></span>
<span><span class="co">## (AAAT)n   616.20410  586.90275  312.54305  369.62323  305.60154  121.613880</span></span>
<span><span class="co">## 625 more rows ...</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="analyze-results-with-edger-1">Analyze results with <em>edgeR</em><a class="anchor" aria-label="anchor" href="#analyze-results-with-edger-1"></a>
</h3>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">atac_ys</span><span class="op">[[</span><span class="st">'repName'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">keep</span>, keep.lib.sizes<span class="op">=</span><span class="cn">TRUE</span><span class="op">]</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html" class="external-link">glmFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">lrt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmLRT.html" class="external-link">glmLRT</a></span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">atac_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">lrt</span>, n<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">lrt</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EnhancedVolcano/man/EnhancedVolcano.html" class="external-link">EnhancedVolcano</a></span><span class="op">(</span><span class="va">atac_res</span>,</span>
<span>                lab<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">atac_res</span><span class="op">)</span>,</span>
<span>                x<span class="op">=</span><span class="st">'logFC'</span>,</span>
<span>                y<span class="op">=</span><span class="st">'PValue'</span>,</span>
<span>                title<span class="op">=</span><span class="st">'Volcano plot for repeat elements'</span>,</span>
<span>                subtitle <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                legendPosition <span class="op">=</span> <span class="st">'right'</span></span>
<span>                <span class="op">)</span></span></code></pre></div>
<p><img src="ratte_files/figure-html/unnamed-chunk-22-1.png" width="576"></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## this is very small test data, we set the cutoff a small number.</span></span>
<span><span class="va">atac_de</span> <span class="op">&lt;-</span> <span class="va">atac_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">atac_res</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">&gt;</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">atac_res</span><span class="op">$</span><span class="va">PValue</span><span class="op">&lt;</span><span class="fl">0.05</span>, <span class="op">]</span></span>
<span><span class="va">plotdata</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">$</span><span class="va">tpm</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">atac_de</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">plotdata</span>, scale <span class="op">=</span> <span class="st">'row'</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="ratte_files/figure-html/unnamed-chunk-22-2.png" width="576"></p>
</div>
</div>
<div class="section level2">
<h2 id="merge-the-results-from-rna-seq-and-atac-seq">Merge the results from RNA-seq and ATAC-seq<a class="anchor" aria-label="anchor" href="#merge-the-results-from-rna-seq-and-atac-seq"></a>
</h2>
<div class="section level3">
<h3 id="for-repeat-elements-level">For repeat elements level<a class="anchor" aria-label="anchor" href="#for-repeat-elements-level"></a>
</h3>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">res</span>, <span class="va">atac_res</span>, suffixes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".RNA"</span>,<span class="st">".ATAC"</span><span class="op">)</span>,</span>
<span>           by<span class="op">=</span><span class="st">'row.names'</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="va">m</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">m</span><span class="op">$</span><span class="va">logFC.RNA</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">m</span><span class="op">$</span><span class="va">logFC.ATAC</span><span class="op">)</span>, </span>
<span>               <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">m</span><span class="op">$</span><span class="va">PValue.RNA</span><span class="op">^</span><span class="fl">2</span> <span class="op">+</span> <span class="va">m</span><span class="op">$</span><span class="va">PValue.ATAC</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">m</span>, n<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      Row.names logFC.RNA logCPM.RNA    LR.RNA   PValue.RNA      FDR.RNA</span></span>
<span><span class="co">## 38  DNA-8-9_DR -1.167023   11.39641  9.773658 1.770299e-03 0.0055523021</span></span>
<span><span class="co">## 28 DNA-1-3B_DR -1.617957   11.15533 16.697501 4.383871e-05 0.0004653648</span></span>
<span><span class="co">##    logFC.ATAC logCPM.ATAC LR.ATAC PValue.ATAC  FDR.ATAC</span></span>
<span><span class="co">## 38 -0.3725652    12.19434 5.48098  0.01922448 0.4170233</span></span>
<span><span class="co">## 28 -0.3158131    12.45271 3.91443  0.04787325 0.5400102</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">seq_anno</span><span class="op">[</span><span class="va">seq_anno</span><span class="op">$</span><span class="va">repName</span><span class="op">==</span><span class="va">m</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span>, <span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 9 columns</span></span>
<span><span class="co">##                  seqname     feature     repName    repClass   repFamily</span></span>
<span><span class="co">##              &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## 1 e7b2400cc9f303582e8d..      intron  DNA-8-9_DR         DNA     hAT-Ac?</span></span>
<span><span class="co">## 2 a3482d5dfc7994ca57dc..      intron  DNA-8-9_DR         DNA     hAT-Ac?</span></span>
<span><span class="co">## 3 b5bc5a06d7fe994c32cf..      intron  DNA-8-9_DR         DNA     hAT-Ac?</span></span>
<span><span class="co">## 4 633c41ea5abc604a0f54..      intron  DNA-8-9_DR         DNA     hAT-Ac?</span></span>
<span><span class="co">## 5 e76f7d6b0d501f6bd997..      intron  DNA-8-9_DR         DNA     hAT-Ac?</span></span>
<span><span class="co">## 6 dab0e247e710af6140ed..      intron  DNA-8-9_DR         DNA     hAT-Ac?</span></span>
<span><span class="co">##           N                    idx                tx_name</span></span>
<span><span class="co">##   &lt;integer&gt;            &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">## 1         1 chr25:14166077-14166..           NM_001128347</span></span>
<span><span class="co">## 2         2 chr25:14251818-14251.. NM_001128347,NM_0011..</span></span>
<span><span class="co">## 3         1 chr25:14879440-14879..              NM_200452</span></span>
<span><span class="co">## 4         1 chr25:11941291-11941..           NM_001256664</span></span>
<span><span class="co">## 5         2 chr25:12061702-12061.. NM_001256664,NM_0012..</span></span>
<span><span class="co">## 6         1 chr25:13092036-13092..           NM_001122750</span></span>
<span><span class="co">##                       seq</span></span>
<span><span class="co">##            &lt;DNAStringSet&gt;</span></span>
<span><span class="co">## 1 GTGCGTACAC...TTTTTGCCAG</span></span>
<span><span class="co">## 2 GTGAGTTGAA...TCTTTCCCAG</span></span>
<span><span class="co">## 3 GTAGGTGTTC...TGATTGGCAG</span></span>
<span><span class="co">## 4 GTAAGCAAGT...CTCTTCGCAG</span></span>
<span><span class="co">## 5 GTGAGTCCTG...TCCCCTGCAG</span></span>
<span><span class="co">## 6 GTAAGAGCTC...TTTTGTGCAG</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">atac_seq_anno</span><span class="op">[</span><span class="va">atac_seq_anno</span><span class="op">$</span><span class="va">repName</span><span class="op">==</span><span class="va">m</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span>, <span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 9 columns</span></span>
<span><span class="co">##                  seqname     feature     repName    repClass   repFamily</span></span>
<span><span class="co">##              &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## 1 4f471f3c7c118eeb7ddb..        peak  DNA-8-9_DR         DNA     hAT-Ac?</span></span>
<span><span class="co">## 2 6a837e21d38fa0082c0f..        peak  DNA-8-9_DR         DNA     hAT-Ac?</span></span>
<span><span class="co">## 3 ff2e591b5fc41b6cff86..        peak  DNA-8-9_DR         DNA     hAT-Ac?</span></span>
<span><span class="co">## 4 0929655c4ade7d453989..        peak  DNA-8-9_DR         DNA     hAT-Ac?</span></span>
<span><span class="co">## 5 b2ec5656f2053898434e..        peak  DNA-8-9_DR         DNA     hAT-Ac?</span></span>
<span><span class="co">## 6 ee081dff4dfbe18cb247..        peak  DNA-8-9_DR         DNA     hAT-Ac?</span></span>
<span><span class="co">##           N                    idx      tx_name                     seq</span></span>
<span><span class="co">##   &lt;integer&gt;            &lt;character&gt;  &lt;character&gt;          &lt;DNAStringSet&gt;</span></span>
<span><span class="co">## 1         1 chr25:11030723-11030.. NM_001265607 AAAATGCTTT...ATTAGTGTAG</span></span>
<span><span class="co">## 2         1 chr25:11324782-11325.. NM_001265607 GGCACTTGTG...ACATATTTAT</span></span>
<span><span class="co">## 3         1 chr25:11358819-11358.. NM_001265607 GATGTCTCCT...GTAGGTCTTA</span></span>
<span><span class="co">## 4         1 chr25:12255981-12256.. NM_001256664 ACTTTCCTTC...GCCAACTGAC</span></span>
<span><span class="co">## 5         1 chr25:12322481-12322..    NM_199670 TTTTTTGTTG...TCGCTGGTTC</span></span>
<span><span class="co">## 6         1 chr25:12322481-12322..    NM_199670 GTTCGCGTGG...GTGTGTGAAT</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="for-transcripts-level">For transcripts level<a class="anchor" aria-label="anchor" href="#for-transcripts-level"></a>
</h3>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DE</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span>  <span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">keep</span>, keep.lib.sizes<span class="op">=</span><span class="cn">TRUE</span><span class="op">]</span></span>
<span>  <span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span>  <span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span>  <span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html" class="external-link">glmFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span>  <span class="va">lrt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmLRT.html" class="external-link">glmLRT</a></span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">lrt</span>, n<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">lrt</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co">## for ATAC-seq/ChIP-seq, the tx_name were annotated by the nearest</span></span>
<span>  <span class="co">## transcript, you may want to change this with other annotation.</span></span>
<span>  <span class="va">tx_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">tx_name</span>, split<span class="op">=</span><span class="st">','</span><span class="op">)</span></span>
<span>  <span class="va">tx_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">tx_name</span>, <span class="va">unique</span><span class="op">)</span></span>
<span>  <span class="va">res</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq.int</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">tx_name</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span>  <span class="va">res</span><span class="op">$</span><span class="va">tx_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">tx_name</span><span class="op">)</span></span>
<span>  <span class="va">res</span></span>
<span><span class="op">}</span></span>
<span><span class="va">tx.RNA</span> <span class="op">&lt;-</span> <span class="fu">DE</span><span class="op">(</span><span class="va">ys</span><span class="op">[[</span><span class="st">'rawCounts'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">tx.ATAC</span> <span class="op">&lt;-</span> <span class="fu">DE</span><span class="op">(</span><span class="va">atac_ys</span><span class="op">[[</span><span class="st">'rawCounts'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tx.RNA</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tx.RNA</span><span class="op">)</span><span class="op">!=</span><span class="st">'seq'</span><span class="op">]</span>, n<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                  feature repName       repClass      repFamily</span></span>
<span><span class="co">## 05fa29a7a02f8c0be64d4a5b26742a72    exon    TDR2            DNA      TcMar-Tc1</span></span>
<span><span class="co">## ffae26ed68dae4a08cfb61df42706275    exon GA-rich Low_complexity Low_complexity</span></span>
<span><span class="co">##                                  N                       idx      tx_name</span></span>
<span><span class="co">## 05fa29a7a02f8c0be64d4a5b26742a72 1 chr25:13715764-13716675:- NM_001324407</span></span>
<span><span class="co">## ffae26ed68dae4a08cfb61df42706275 1 chr25:13525201-13525231:+    NM_200833</span></span>
<span><span class="co">##                                      logFC   logCPM       LR       PValue</span></span>
<span><span class="co">## 05fa29a7a02f8c0be64d4a5b26742a72 -3.495283 13.62574 14.21923 0.0001626991</span></span>
<span><span class="co">## ffae26ed68dae4a08cfb61df42706275  2.240721 17.91356 14.19749 0.0001645899</span></span>
<span><span class="co">##                                          FDR</span></span>
<span><span class="co">## 05fa29a7a02f8c0be64d4a5b26742a72 0.003620978</span></span>
<span><span class="co">## ffae26ed68dae4a08cfb61df42706275 0.003620978</span></span></code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tx.ATAC</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tx.ATAC</span><span class="op">)</span><span class="op">!=</span><span class="st">'seq'</span><span class="op">]</span>, n<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                  feature    repName repClass repFamily N</span></span>
<span><span class="co">## 7f8ac9a6f193c08adc00e8ffd48e318b    peak DNA-8-9_DR      DNA   hAT-Ac? 1</span></span>
<span><span class="co">## 10187e7a0d6b8c9a71d0986072adf236    peak    HAT1_DR      DNA    hAT-Ac 2</span></span>
<span><span class="co">##                                                                                  idx</span></span>
<span><span class="co">## 7f8ac9a6f193c08adc00e8ffd48e318b                           chr25:14166077-14166094:+</span></span>
<span><span class="co">## 10187e7a0d6b8c9a71d0986072adf236 chr25:12652401-12652609:+,chr25:12652706-12652799:+</span></span>
<span><span class="co">##                                       tx_name     logFC   logCPM       LR</span></span>
<span><span class="co">## 7f8ac9a6f193c08adc00e8ffd48e318b NM_001128347 -1.024255 12.47134 10.59651</span></span>
<span><span class="co">## 10187e7a0d6b8c9a71d0986072adf236 NM_001122702  1.333532 10.66241 10.31417</span></span>
<span><span class="co">##                                       PValue       FDR</span></span>
<span><span class="co">## 7f8ac9a6f193c08adc00e8ffd48e318b 0.001133015 0.2026396</span></span>
<span><span class="co">## 10187e7a0d6b8c9a71d0986072adf236 0.001320128 0.2026396</span></span></code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tx_level</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">tx.RNA</span>, <span class="va">tx.ATAC</span>,</span>
<span>                  by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'repName'</span>, <span class="st">'repClass'</span>, <span class="st">'repFamily'</span>, <span class="st">'tx_name'</span><span class="op">)</span>,</span>
<span>                  suffix<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'.RNA'</span>, <span class="st">'.ATAC'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tx_level</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'tx_name'</span>, <span class="st">'repName'</span>, <span class="st">'feature.RNA'</span>,</span>
<span>                <span class="st">'logFC.RNA'</span>, <span class="st">'PValue.RNA'</span>, <span class="st">'FDR.RNA'</span>,</span>
<span>                <span class="st">'logFC.ATAC'</span>, <span class="st">'PValue.ATAC'</span>, <span class="st">'FDR.ATAC'</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##        tx_name     repName feature.RNA  logFC.RNA PValue.RNA   FDR.RNA</span></span>
<span><span class="co">## 1 NM_001003460 DNA-1-11_DR      intron -0.4502847 0.67752822 0.8517498</span></span>
<span><span class="co">## 2    NM_200042  DNA-4-3_DR      intron -1.7903609 0.02994221 0.1209145</span></span>
<span><span class="co">##   logFC.ATAC PValue.ATAC  FDR.ATAC</span></span>
<span><span class="co">## 1 0.02855417   0.9578389 0.9900239</span></span>
<span><span class="co">## 2 0.50479076   0.1850781 0.9522626</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] pheatmap_1.0.13                        </span></span>
<span><span class="co">##  [2] EnhancedVolcano_1.27.0                 </span></span>
<span><span class="co">##  [3] ggrepel_0.9.6                          </span></span>
<span><span class="co">##  [4] ggplot2_3.5.2                          </span></span>
<span><span class="co">##  [5] TxDb.Drerio.UCSC.danRer11.refGene_3.4.6</span></span>
<span><span class="co">##  [6] GenomicFeatures_1.61.4                 </span></span>
<span><span class="co">##  [7] AnnotationDbi_1.71.0                   </span></span>
<span><span class="co">##  [8] Biobase_2.69.0                         </span></span>
<span><span class="co">##  [9] BSgenome.Drerio.UCSC.danRer11_1.4.2    </span></span>
<span><span class="co">## [10] BSgenome_1.77.1                        </span></span>
<span><span class="co">## [11] BiocIO_1.19.0                          </span></span>
<span><span class="co">## [12] Biostrings_2.77.2                      </span></span>
<span><span class="co">## [13] XVector_0.49.0                         </span></span>
<span><span class="co">## [14] AnnotationHub_3.99.6                   </span></span>
<span><span class="co">## [15] BiocFileCache_2.99.5                   </span></span>
<span><span class="co">## [16] dbplyr_2.5.0                           </span></span>
<span><span class="co">## [17] edgeR_4.7.2                            </span></span>
<span><span class="co">## [18] limma_3.65.1                           </span></span>
<span><span class="co">## [19] rtracklayer_1.69.1                     </span></span>
<span><span class="co">## [20] GenomicRanges_1.61.1                   </span></span>
<span><span class="co">## [21] Seqinfo_0.99.1                         </span></span>
<span><span class="co">## [22] IRanges_2.43.0                         </span></span>
<span><span class="co">## [23] S4Vectors_0.47.0                       </span></span>
<span><span class="co">## [24] BiocGenerics_0.55.0                    </span></span>
<span><span class="co">## [25] generics_0.1.4                         </span></span>
<span><span class="co">## [26] ratte_0.0.6                            </span></span>
<span><span class="co">## [27] BiocStyle_2.37.0                       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] DBI_1.2.3                   bitops_1.0-9               </span></span>
<span><span class="co">##  [3] httr2_1.1.2                 rlang_1.1.6                </span></span>
<span><span class="co">##  [5] magrittr_2.0.3              matrixStats_1.5.0          </span></span>
<span><span class="co">##  [7] compiler_4.5.1              RSQLite_2.4.1              </span></span>
<span><span class="co">##  [9] png_0.1-8                   systemfonts_1.2.3          </span></span>
<span><span class="co">## [11] vctrs_0.6.5                 pkgconfig_2.0.3            </span></span>
<span><span class="co">## [13] crayon_1.5.3                fastmap_1.2.0              </span></span>
<span><span class="co">## [15] labeling_0.4.3              Rsamtools_2.25.1           </span></span>
<span><span class="co">## [17] rmarkdown_2.29              tzdb_0.5.0                 </span></span>
<span><span class="co">## [19] UCSC.utils_1.5.0            ragg_1.4.0                 </span></span>
<span><span class="co">## [21] purrr_1.0.4                 bit_4.6.0                  </span></span>
<span><span class="co">## [23] xfun_0.52                   cachem_1.1.0               </span></span>
<span><span class="co">## [25] GenomeInfoDb_1.45.6         jsonlite_2.0.0             </span></span>
<span><span class="co">## [27] blob_1.2.4                  DelayedArray_0.35.2        </span></span>
<span><span class="co">## [29] BiocParallel_1.43.4         parallel_4.5.1             </span></span>
<span><span class="co">## [31] R6_2.6.1                    bslib_0.9.0                </span></span>
<span><span class="co">## [33] RColorBrewer_1.1-3          jquerylib_0.1.4            </span></span>
<span><span class="co">## [35] Rcpp_1.0.14                 bookdown_0.43              </span></span>
<span><span class="co">## [37] SummarizedExperiment_1.39.1 knitr_1.50                 </span></span>
<span><span class="co">## [39] readr_2.1.5                 splines_4.5.1              </span></span>
<span><span class="co">## [41] Matrix_1.7-3                tidyselect_1.2.1           </span></span>
<span><span class="co">## [43] abind_1.4-8                 yaml_2.3.10                </span></span>
<span><span class="co">## [45] codetools_0.2-20            curl_6.4.0                 </span></span>
<span><span class="co">## [47] lattice_0.22-7              tibble_3.3.0               </span></span>
<span><span class="co">## [49] withr_3.0.2                 KEGGREST_1.49.1            </span></span>
<span><span class="co">## [51] evaluate_1.0.4              desc_1.4.3                 </span></span>
<span><span class="co">## [53] pillar_1.10.2               BiocManager_1.30.26        </span></span>
<span><span class="co">## [55] filelock_1.0.3              MatrixGenerics_1.21.0      </span></span>
<span><span class="co">## [57] RCurl_1.98-1.17             BiocVersion_3.22.0         </span></span>
<span><span class="co">## [59] hms_1.1.3                   scales_1.4.0               </span></span>
<span><span class="co">## [61] glue_1.8.0                  tools_4.5.1                </span></span>
<span><span class="co">## [63] data.table_1.17.6           locfit_1.5-9.12            </span></span>
<span><span class="co">## [65] GenomicAlignments_1.45.1    fs_1.6.6                   </span></span>
<span><span class="co">## [67] XML_3.99-0.18               grid_4.5.1                 </span></span>
<span><span class="co">## [69] restfulr_0.0.15             cli_3.6.5                  </span></span>
<span><span class="co">## [71] rappdirs_0.3.3              textshaping_1.0.1          </span></span>
<span><span class="co">## [73] S4Arrays_1.9.1              dplyr_1.1.4                </span></span>
<span><span class="co">## [75] gtable_0.3.6                sass_0.4.10                </span></span>
<span><span class="co">## [77] digest_0.6.37               SparseArray_1.9.0          </span></span>
<span><span class="co">## [79] rjson_0.2.23                htmlwidgets_1.6.4          </span></span>
<span><span class="co">## [81] farver_2.1.2                memoise_2.0.1              </span></span>
<span><span class="co">## [83] htmltools_0.5.8.1           pkgdown_2.1.3              </span></span>
<span><span class="co">## [85] lifecycle_1.0.4             httr_1.4.7                 </span></span>
<span><span class="co">## [87] statmod_1.5.0               bit64_4.6.0-1</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body" entry-spacing="0" line-spacing="2">
<div id="ref-kong2019transposable" class="csl-entry">
<div class="csl-left-margin">1. </div>
<div class="csl-right-inline">Kong, Y. <em>et al.</em> Transposable element
expression in tumors is associated with immune infiltration and
increased antigenicity. <em>Nature communications</em>
<strong>10,</strong> 5228 (2019).</div>
</div>
<div id="ref-thompson2020identification" class="csl-entry">
<div class="csl-left-margin">2. </div>
<div class="csl-right-inline">Thompson, J. D. <em>et al.</em> Identification
and requirements of enhancers that direct gene expression during
zebrafish fin regeneration. <em>Development</em> <strong>147,</strong>
dev191262 (2020).</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
