<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ratte Guide • ratte</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ratte Guide">
<meta property="og:description" content="ratte">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ratte</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/ratte.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jianhong/ratte" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>ratte Guide</h1>
                        <h4 data-toc-skip class="author">Jianhong
Ou</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jianhong/ratte/blob/HEAD/vignettes/ratte.Rmd" class="external-link"><code>vignettes/ratte.Rmd</code></a></small>
      <div class="hidden name"><code>ratte.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="section-introduction">Introduction<a class="anchor" aria-label="anchor" href="#section-introduction"></a>
</h2>
<p>Transposable elements (TEs) play a role in fostering genetic
diversity, instigating genome rearrangements, and exerting both positive
and negative effects on the host organism. The regulation of these
elements constitutes a dynamic and intricate process, involving host
defense mechanisms to control their activity. The <a href="https://github.com/ucsffrancislab/REdiscoverTE" class="external-link">REdiscoverTE</a>,
presented in 2019 by Kong et al.<span class="citation"><sup>1</sup></span>, showcased remarkable accuracy in
TE quantification. It stands out as a significantly more comprehensive
and accurate method for robust, whole-transcriptome quantification.</p>
<p>Nevertheless, the existing <em>REdiscoverTE</em> is constrained by
the human genome assembly GRCh38. To broaden the scope of annotation to
encompass more species, the <em>ratte</em> package has been devised.
This package leverages the resources available within the Bioconductor
community to prepare the necessary sequence files and annotations.</p>
</div>
<div class="section level2">
<h2 id="section-quick-start-for-rna-seq-data">Quick start for RNA-seq data<a class="anchor" aria-label="anchor" href="#section-quick-start-for-rna-seq-data"></a>
</h2>
<p>Here is an example using <em>ratte</em> with a subset of RNA-seq
data.</p>
<div class="section level3">
<h3 id="section-installation">Installation<a class="anchor" aria-label="anchor" href="#section-installation"></a>
</h3>
<p>First, install <em>ratte</em> and other packages required to run the
examples. Please note that the example dataset used here is from
zebrafish. To run analysis with dataset from a different species or
different assembly, please install the corresponding Bsgenome and TxDb.
For example, to analyze mouse data aligned to mm10, please install
<code>BSgenome.Mmusculus.UCSC.mm10</code>, and
<code>TxDb.Mmusculus.UCSC.mm10.knownGene</code>. You can also generate a
TxDb object by functions <code>makeTxDbFromGFF</code> from a local gff
file, or <code>makeTxDbFromUCSC</code>,
<code>makeTxDbFromBiomart</code>, and <code>makeTxDbFromEnsembl</code>,
from online resources in <em>GenomicFeatures</em> package. The BSgenome
package can be created by <em>BSgenomeForge</em> package.</p>
<p>Please note that the example dataset used here is from a small subset
of zebrafish fin regeneration<span class="citation"><sup>2</sup></span>.</p>
<div class="sourceCode" id="section-cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"jianhong/ratte"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## other annotation and helper libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">BiocManager</a></span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AnnotationHub"</span>, <span class="st">"rtracklayer"</span>,</span>
<span>                       <span class="st">"BSgenome.Drerio.UCSC.danRer11"</span>,</span>
<span>                       <span class="st">"TxDb.Drerio.UCSC.danRer11.refGene"</span>,</span>
<span>                       <span class="st">"EnhancedVolcano"</span>,</span>
<span>                       <span class="st">"pheatmap"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="section-load-library">Load library<a class="anchor" aria-label="anchor" href="#section-load-library"></a>
</h3>
<div class="sourceCode" id="section-cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jianhong/ratte" class="external-link">ratte</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="section-generating-reference-transcriptome">Generating reference transcriptome<a class="anchor" aria-label="anchor" href="#section-generating-reference-transcriptome"></a>
</h3>
<p>The reference transcriptome includes:</p>
<ol style="list-style-type: decimal">
<li><p>Distinct RNA transcript sequences prepared by
<code><a href="https://rdrr.io/pkg/GenomicFeatures/man/extractTranscriptSeqs.html" class="external-link">GenomicFeatures::extractTranscriptSeqs</a></code> function for output
of <code><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html" class="external-link">GenomicFeatures::exonsBy</a></code> at transcript level.</p></li>
<li><p>Distinct sequences representing RE-containing introns by
<code><a href="https://rdrr.io/pkg/GenomicFeatures/man/exonicParts.html" class="external-link">GenomicFeatures::intronicParts</a></code>.</p></li>
<li><p>Distinct RepeatMasker elements in the intergenic region.</p></li>
</ol>
<p>In this manual, we will use the fish genome.</p>
<div class="sourceCode" id="section-cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Drerio.UCSC.danRer11</span><span class="op">)</span></span>
<span><span class="co">## set genome, Drerio is a shortname for BSgenome.Drerio.UCSC.danRer11</span></span>
<span><span class="va">genome</span> <span class="op">&lt;-</span> <span class="va">Drerio</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Drerio.UCSC.danRer11.refGene</span><span class="op">)</span></span>
<span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Drerio.UCSC.danRer11.refGene</span> <span class="co">## give it a short name</span></span></code></pre></div>
<div class="sourceCode" id="section-cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ah</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">rmsk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">ah</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'RepeatMasker'</span>, <span class="st">'Zebrafish'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## AnnotationHub with 1 record</span></span>
<span><span class="co">## # snapshotDate(): 2023-11-08</span></span>
<span><span class="co">## # names(): AH98980</span></span>
<span><span class="co">## # $dataprovider: UCSC</span></span>
<span><span class="co">## # $species: Danio rerio</span></span>
<span><span class="co">## # $rdataclass: GRanges</span></span>
<span><span class="co">## # $rdatadateadded: 2022-01-31</span></span>
<span><span class="co">## # $title: UCSC RepeatMasker annotations (Nov2017) for Zebrafish (danRer11)</span></span>
<span><span class="co">## # $description: UCSC RepeatMasker annotations (Nov2017) for Zebrafish -Danio...</span></span>
<span><span class="co">## # $taxonomyid: 7955</span></span>
<span><span class="co">## # $genome: danRer11</span></span>
<span><span class="co">## # $sourcetype: UCSC track</span></span>
<span><span class="co">## # $sourceurl: https://functionalgenomics.upf.edu/annotationhub/repeatmasker/...</span></span>
<span><span class="co">## # $sourcesize: NA</span></span>
<span><span class="co">## # $tags: c("RepeatMasker", "UCSC") </span></span>
<span><span class="co">## # retrieve record with 'object[["AH98980"]]'</span></span></code></pre>
<div class="sourceCode" id="section-cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rmsk</span> <span class="op">&lt;-</span> <span class="va">rmsk</span><span class="op">[[</span><span class="st">'AH98980'</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>We will use a subset of fish genome in this manual. Please skip this
step if you want try the full fish genome.</p>
<div class="sourceCode" id="section-cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rmsk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'danRer11_rmsk_chr25_sub.RDS'</span>,</span>
<span>                            package<span class="op">=</span><span class="st">'ratte'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The <code>prepareSeq</code> function will extract the sequences and
merge the RepeatMasker elements annotation. The <code>saveFasta</code>
function is uesed to export the sequences into a fasta file.</p>
<div class="sourceCode" id="section-cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareSeq.html">prepareSeq</a></span><span class="op">(</span><span class="va">genome</span>, <span class="va">txdb</span>, <span class="va">rmsk</span>,</span>
<span><span class="co">## The subsetGRanges parameter is only used for subset the genome.</span></span>
<span><span class="co">## Please remove this parameter for full fish genome.</span></span>
<span>                       subsetGRanges<span class="op">=</span><span class="fu">GRanges</span><span class="op">(</span><span class="st">'chr25:11000000-15000000'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rna_outfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext<span class="op">=</span><span class="st">'.fa.gz'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/saveFasta.html">saveFasta</a></span><span class="op">(</span><span class="va">seq_anno</span>, <span class="va">rna_outfile</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="section-align-reads-by-salmon">Align reads by salmon<a class="anchor" aria-label="anchor" href="#section-align-reads-by-salmon"></a>
</h3>
<p>First we will create salmon index. In this example, we will just
export the command by setting <code>dryrun</code> parameter to TRUE.</p>
<div class="sourceCode" id="section-cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prepareSalmonIndex.html">prepareSalmonIndex</a></span><span class="op">(</span>salmonPath<span class="op">=</span><span class="st">'salmon'</span>,</span>
<span>                   FaFile<span class="op">=</span><span class="va">rna_outfile</span>,</span>
<span>                   index<span class="op">=</span><span class="st">'RNA_TEindex'</span>,</span>
<span>                   dryrun<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## salmon index --threads 2 -t /tmp/RtmpVHRAQO/file1285f0a120f.fa.gz -i RNA_TEindex</span></span></code></pre>
<p>Second, we will align reads to salmon index.</p>
<div class="sourceCode" id="section-cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/alignReadsBySalmon.html">alignReadsBySalmon</a></span><span class="op">(</span><span class="st">'sample1.R1.fq.gz'</span>, <span class="st">'sample1.R2.fq.gz'</span>,</span>
<span>                    output<span class="op">=</span><span class="st">'sample1_salmon_output'</span>,</span>
<span>                    index<span class="op">=</span><span class="st">'RNA_TEindex'</span>,</span>
<span>                    dryrun<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## salmon quant --seqBias --gcBias --index RNA_TEindex --libType A --validateMappings --threads 2 -o sample1_salmon_output -1 sample1.R1.fq.gz -2 sample1.R2.fq.gz</span></span></code></pre>
<p>Once alignment is done, we will get quantification file named as
<code>quant.sf</code>. In this manual, we will load pre-aligned
files.</p>
<div class="sourceCode" id="section-cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">quant_sfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'RNA-seq'</span>, package<span class="op">=</span><span class="st">'ratte'</span><span class="op">)</span>, <span class="st">'quant.sf'</span>,</span>
<span>                 full.names<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">qnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"^.*\\/(.*?).quant.sf"</span>, <span class="st">"\\1"</span>, <span class="va">quant_sfs</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="section-aggregate-alignments-to-repeat-element-re-">Aggregate alignments to repeat element (RE).<a class="anchor" aria-label="anchor" href="#section-aggregate-alignments-to-repeat-element-re-"></a>
</h3>
<p>The <code>createDGELists</code> function will extract counts number
from the <code>quant.sf</code> files and generate a list of
<code><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">edgeR::DGEList</a></code> object. The available choices for
<code>select_feature</code> are ‘all’, ‘exon’, ‘intron’, ‘gene’, and
‘intergenic’. Here we use <code>all</code> as an example. Please note
that the DGEList library size and normliazation factor will be
determined by gene level counts.</p>
<div class="sourceCode" id="section-cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ys</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createDGELists.html">createDGELists</a></span><span class="op">(</span><span class="va">quant_sfs</span>, <span class="va">qnames</span>, <span class="va">seq_anno</span>,</span>
<span>                     select_feature <span class="op">=</span> <span class="st">'all'</span>,</span>
<span>                     group<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">'.rep.'</span>, <span class="st">''</span>, <span class="va">qnames</span><span class="op">)</span>,</span>
<span>                     remove.zeros <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Removing 447 rows with all zero counts</span></span></code></pre>
<pre><code><span><span class="co">## Removing 2 rows with all zero counts</span></span></code></pre>
<pre><code><span><span class="co">## Removing 6 rows with all zero counts</span></span></code></pre>
<div class="sourceCode" id="section-cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ys</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "repName"   "repClass"  "repFamily"</span></span></code></pre>
<div class="sourceCode" id="section-cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ys</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "DGEList"</span></span>
<span><span class="co">## $counts</span></span>
<span><span class="co">##           Reg.rep1 Reg.rep2 Uni.rep1 Uni.rep2</span></span>
<span><span class="co">## (A)n        22.143   27.214   28.546   19.000</span></span>
<span><span class="co">## (AACTGA)n   11.239   17.000   11.000    4.000</span></span>
<span><span class="co">## (AAT)n      11.000    7.000   47.000   46.000</span></span>
<span><span class="co">## (AATA)n      8.006    8.000   15.529    7.001</span></span>
<span><span class="co">## (AATAA)n     3.000    1.000    4.000    3.000</span></span>
<span><span class="co">## 456 more rows ...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $samples</span></span>
<span><span class="co">##          group  lib.size norm.factors</span></span>
<span><span class="co">## Reg.rep1   Reg  7605.759            1</span></span>
<span><span class="co">## Reg.rep2   Reg  8182.806            1</span></span>
<span><span class="co">## Uni.rep1   Uni 13468.979            1</span></span>
<span><span class="co">## Uni.rep2   Uni 12068.875            1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $tpm</span></span>
<span><span class="co">##            Reg.rep1  Reg.rep2   Uni.rep1   Uni.rep2</span></span>
<span><span class="co">## (A)n      18.625446 19.844470 175.139452 160.457251</span></span>
<span><span class="co">## (AACTGA)n  5.391495  6.781941   5.878126   2.179567</span></span>
<span><span class="co">## (AAT)n     9.644679 11.832648  63.671101  68.940900</span></span>
<span><span class="co">## (AATA)n   49.648589 46.288508 119.055364  94.325014</span></span>
<span><span class="co">## (AATAA)n   3.411824  0.443921   3.333512   3.882252</span></span>
<span><span class="co">## 456 more rows ...</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="section-analyze-results-with-edger">Analyze results with <em>edgeR</em><a class="anchor" aria-label="anchor" href="#section-analyze-results-with-edger"></a>
</h3>
<div class="sourceCode" id="section-cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kevinblighe/EnhancedVolcano" class="external-link">EnhancedVolcano</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">ys</span><span class="op">[[</span><span class="st">'repName'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">keep</span>, keep.lib.sizes<span class="op">=</span><span class="cn">TRUE</span><span class="op">]</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html" class="external-link">glmFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">lrt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html" class="external-link">glmLRT</a></span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">lrt</span>, n<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">lrt</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EnhancedVolcano/man/EnhancedVolcano.html" class="external-link">EnhancedVolcano</a></span><span class="op">(</span><span class="va">res</span>,</span>
<span>                lab<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>,</span>
<span>                x<span class="op">=</span><span class="st">'logFC'</span>,</span>
<span>                y<span class="op">=</span><span class="st">'PValue'</span>,</span>
<span>                title<span class="op">=</span><span class="st">'Volcano plot for repeat elements'</span>,</span>
<span>                subtitle <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                legendPosition <span class="op">=</span> <span class="st">'right'</span></span>
<span>                <span class="op">)</span></span></code></pre></div>
<p><img src="ratte_files/figure-html/unnamed-chunk-13-1.png" width="576"></p>
<div class="sourceCode" id="section-cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">2</span> <span class="op">&amp;</span> <span class="va">res</span><span class="op">$</span><span class="va">FDR</span><span class="op">&lt;</span><span class="fl">0.05</span>, <span class="op">]</span></span>
<span><span class="va">plotdata</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">$</span><span class="va">tpm</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">de</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">plotdata</span>, scale <span class="op">=</span> <span class="st">'row'</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="ratte_files/figure-html/unnamed-chunk-13-2.png" width="576"></p>
</div>
</div>
<div class="section level2">
<h2 id="section-quick-start-for-atac-seq-data">Quick start for ATAC-seq data<a class="anchor" aria-label="anchor" href="#section-quick-start-for-atac-seq-data"></a>
</h2>
<div class="section level3">
<h3 id="section-generating-reference-for-atac-seq">Generating reference for ATAC-seq<a class="anchor" aria-label="anchor" href="#section-generating-reference-for-atac-seq"></a>
</h3>
<p>The reference includes:</p>
<ol style="list-style-type: decimal">
<li><p>Distinct exon sequences prepared by <code><a href="https://rdrr.io/pkg/Biostrings/man/getSeq.html" class="external-link">BSgenome::getSeq</a></code>
function for output of <code><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">GenomicFeatures::exons</a></code>.</p></li>
<li><p>Distinct sequences representing RE-containing introns by
<code><a href="https://rdrr.io/pkg/GenomicFeatures/man/exonicParts.html" class="external-link">GenomicFeatures::intronicParts</a></code>.</p></li>
<li><p>Distinct RepeatMasker elements in the intergenic region.</p></li>
</ol>
<p>For ATAC-seq, to prepare the sequences, the called peaks are
required.</p>
<div class="sourceCode" id="section-cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'merged.peaks.bed.gz'</span>,</span>
<span>                            package <span class="op">=</span> <span class="st">'ratte'</span><span class="op">)</span>,</span>
<span>                format <span class="op">=</span> <span class="st">'bed'</span><span class="op">)</span></span>
<span><span class="va">atac_seq_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareSeq.html">prepareSeq</a></span><span class="op">(</span><span class="va">genome</span>, <span class="va">txdb</span>, <span class="va">rmsk</span>, <span class="va">peaks</span><span class="op">)</span></span>
<span><span class="va">atac_outfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext<span class="op">=</span><span class="st">'.fa.gz'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/saveFasta.html">saveFasta</a></span><span class="op">(</span><span class="va">atac_seq_anno</span>, <span class="va">atac_outfile</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="section-align-reads-by-salmon-1">Align reads by salmon<a class="anchor" aria-label="anchor" href="#section-align-reads-by-salmon-1"></a>
</h3>
<p>First we will create salmon index. In this example, we will just
export the command by setting <code>dryrun</code> parameter to TRUE.</p>
<div class="sourceCode" id="section-cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prepareSalmonIndex.html">prepareSalmonIndex</a></span><span class="op">(</span>salmonPath<span class="op">=</span><span class="st">'salmon'</span>,</span>
<span>                   FaFile<span class="op">=</span><span class="va">atac_outfile</span>,</span>
<span>                   index<span class="op">=</span><span class="st">'ATAC_TEindex'</span>,</span>
<span>                   dryrun<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## salmon index --threads 2 -t /tmp/RtmpVHRAQO/file128551ad5455.fa.gz -i ATAC_TEindex</span></span></code></pre>
<p>Second, we will align reads to salmon index.</p>
<div class="sourceCode" id="section-cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/alignReadsBySalmon.html">alignReadsBySalmon</a></span><span class="op">(</span><span class="st">'sample1.R1.fq.gz'</span>, <span class="st">'sample1.R2.fq.gz'</span>,</span>
<span>                    output<span class="op">=</span><span class="st">'sample1_salmon_output'</span>,</span>
<span>                    index<span class="op">=</span><span class="st">'ATAC_TEindex'</span>,</span>
<span>                    dryrun<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## salmon quant --seqBias --gcBias --index ATAC_TEindex --libType A --validateMappings --threads 2 -o sample1_salmon_output -1 sample1.R1.fq.gz -2 sample1.R2.fq.gz</span></span></code></pre>
<p>Once alignment is done, we will get quantification file named as
<code>quant.sf</code>. In this manual, we will load pre-aligned
files.</p>
<div class="sourceCode" id="section-cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">atac_quant_sfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'ATAC-seq'</span>,</span>
<span>                                  package<span class="op">=</span><span class="st">'ratte'</span><span class="op">)</span>,</span>
<span>                      <span class="st">'quant.sf'</span>,</span>
<span>                      full.names<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">atac_qnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"^.*\\/(.*?).quant.sf"</span>, <span class="st">"\\1"</span>, <span class="va">atac_quant_sfs</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="section-aggregate-alignments-to-repeat-element-re--1">Aggregate alignments to repeat element (RE).<a class="anchor" aria-label="anchor" href="#section-aggregate-alignments-to-repeat-element-re--1"></a>
</h3>
<p>The <code>createDGELists</code> function will extract counts number
from the <code>quant.sf</code> files and generate a list of
<code><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">edgeR::DGEList</a></code> object. The available choices for
<code>select_feature</code> are ‘all’, ‘exon’, ‘intron’, ‘gene’, and
‘intergenic’. Here we use <code>all</code> as an example. Please note
that different from RNA-seq data, the DGEList library size and
normliazation factor will be determined by <code>all</code> feature
counts.</p>
<div class="sourceCode" id="section-cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">atac_ys</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createDGELists.html">createDGELists</a></span><span class="op">(</span><span class="va">atac_quant_sfs</span>, <span class="va">atac_qnames</span>, <span class="va">atac_seq_anno</span>,</span>
<span>                          select_feature <span class="op">=</span> <span class="st">'all'</span>,</span>
<span>                          group<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">'.rep.'</span>, <span class="st">''</span>, <span class="va">atac_qnames</span><span class="op">)</span>,</span>
<span>                          norm_level<span class="op">=</span><span class="st">'all'</span>,</span>
<span>                          remove.zeros <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Removing 139 rows with all zero counts</span></span></code></pre>
<pre><code><span><span class="co">## Removing 1 rows with all zero counts</span></span></code></pre>
<div class="sourceCode" id="section-cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">atac_ys</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "repName"   "repClass"  "repFamily"</span></span></code></pre>
<div class="sourceCode" id="section-cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">atac_ys</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "DGEList"</span></span>
<span><span class="co">## $counts</span></span>
<span><span class="co">##           Reg.rep1 Reg.rep2 Uni.rep1 Uni.rep2</span></span>
<span><span class="co">## (A)n       758.042  822.216  646.107  660.581</span></span>
<span><span class="co">## (AAAT)n      0.000    1.000    0.000    1.000</span></span>
<span><span class="co">## (AAC)n       0.000    0.989    0.000    0.000</span></span>
<span><span class="co">## (AACTGA)n  342.003  383.003  324.614  334.387</span></span>
<span><span class="co">## (AAT)n     408.034  331.100  304.965  320.125</span></span>
<span><span class="co">## 575 more rows ...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $samples</span></span>
<span><span class="co">##          group lib.size norm.factors</span></span>
<span><span class="co">## Reg.rep1   Reg 248858.0            1</span></span>
<span><span class="co">## Reg.rep2   Reg 262879.2            1</span></span>
<span><span class="co">## Uni.rep1   Uni 231958.5            1</span></span>
<span><span class="co">## Uni.rep2   Uni 238158.1            1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $tpm</span></span>
<span><span class="co">##            Reg.rep1   Reg.rep2  Uni.rep1  Uni.rep2</span></span>
<span><span class="co">## (A)n      547.01449  753.04628 828.20010 813.54697</span></span>
<span><span class="co">## (AAAT)n     0.00000  128.19506   0.00000 135.77487</span></span>
<span><span class="co">## (AAC)n      0.00000 1901.43632   0.00000   0.00000</span></span>
<span><span class="co">## (AACTGA)n  94.26064   99.61734  85.43542  91.94861</span></span>
<span><span class="co">## (AAT)n    375.63789 2240.41052 286.08757 306.64686</span></span>
<span><span class="co">## 575 more rows ...</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="section-analyze-results-with-edger-1">Analyze results with <em>edgeR</em><a class="anchor" aria-label="anchor" href="#section-analyze-results-with-edger-1"></a>
</h3>
<div class="sourceCode" id="section-cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">atac_ys</span><span class="op">[[</span><span class="st">'repName'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">keep</span>, keep.lib.sizes<span class="op">=</span><span class="cn">TRUE</span><span class="op">]</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html" class="external-link">glmFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">lrt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html" class="external-link">glmLRT</a></span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">atac_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">lrt</span>, n<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">lrt</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EnhancedVolcano/man/EnhancedVolcano.html" class="external-link">EnhancedVolcano</a></span><span class="op">(</span><span class="va">atac_res</span>,</span>
<span>                lab<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">atac_res</span><span class="op">)</span>,</span>
<span>                x<span class="op">=</span><span class="st">'logFC'</span>,</span>
<span>                y<span class="op">=</span><span class="st">'PValue'</span>,</span>
<span>                title<span class="op">=</span><span class="st">'Volcano plot for repeat elements'</span>,</span>
<span>                subtitle <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                legendPosition <span class="op">=</span> <span class="st">'right'</span></span>
<span>                <span class="op">)</span></span></code></pre></div>
<p><img src="ratte_files/figure-html/unnamed-chunk-19-1.png" width="576"></p>
<div class="sourceCode" id="section-cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">atac_de</span> <span class="op">&lt;-</span> <span class="va">atac_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">atac_res</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">&gt;</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">atac_res</span><span class="op">$</span><span class="va">PValue</span><span class="op">&lt;</span><span class="fl">0.05</span>, <span class="op">]</span></span>
<span><span class="va">plotdata</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">$</span><span class="va">tpm</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">atac_de</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">plotdata</span>, scale <span class="op">=</span> <span class="st">'row'</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="ratte_files/figure-html/unnamed-chunk-19-2.png" width="576"></p>
</div>
</div>
<div class="section level2">
<h2 id="section-merge-the-results-from-rna-seq-and-atac-seq">Merge the results from RNA-seq and ATAC-seq<a class="anchor" aria-label="anchor" href="#section-merge-the-results-from-rna-seq-and-atac-seq"></a>
</h2>
<div class="sourceCode" id="section-cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">res</span>, <span class="va">atac_res</span>, suffixes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".RNA"</span>,<span class="st">".ATAC"</span><span class="op">)</span>,</span>
<span>           by<span class="op">=</span><span class="st">'row.names'</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="va">m</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">m</span><span class="op">$</span><span class="va">logFC.RNA</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">m</span><span class="op">$</span><span class="va">logFC.ATAC</span><span class="op">)</span>, </span>
<span>               <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">m</span><span class="op">$</span><span class="va">PValue.RNA</span><span class="op">^</span><span class="fl">2</span> <span class="op">+</span> <span class="va">m</span><span class="op">$</span><span class="va">PValue.ATAC</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">m</span>, n<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         Row.names logFC.RNA logCPM.RNA   LR.RNA  PValue.RNA    FDR.RNA</span></span>
<span><span class="co">## 54     DIRS-15_DR -1.260937   10.65210 7.941606 0.004831094 0.01514989</span></span>
<span><span class="co">## 144 Mariner-N5_DR -1.281433   10.82231 9.597140 0.001948807 0.01053574</span></span>
<span><span class="co">##     logFC.ATAC logCPM.ATAC  LR.ATAC PValue.ATAC   FDR.ATAC</span></span>
<span><span class="co">## 54  -0.2632025     9.99593 8.000676 0.004675988 0.09713198</span></span>
<span><span class="co">## 144 -0.1369858    11.75537 7.019590 0.008062267 0.09713198</span></span></code></pre>
<div class="sourceCode" id="section-cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">seq_anno</span><span class="op">[</span><span class="va">seq_anno</span><span class="op">$</span><span class="va">repName</span><span class="op">==</span><span class="va">m</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span>, <span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 8 columns</span></span>
<span><span class="co">##                  seqname     feature     repName    repClass   repFamily</span></span>
<span><span class="co">##              &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## 1 6f9e6fe1e4efb369c29c..      intron  DIRS-15_DR         LTR       Ngaro</span></span>
<span><span class="co">## 2 ff1ddbaa96922149aeb3..  intergenic  DIRS-15_DR         LTR       Ngaro</span></span>
<span><span class="co">## 3 7c6a1e4430f8eed7a669..  intergenic  DIRS-15_DR         LTR       Ngaro</span></span>
<span><span class="co">## 4 e78ec63e8c31d44a9387..  intergenic  DIRS-15_DR         LTR       Ngaro</span></span>
<span><span class="co">## 5 da2b9f4bb6f238c3c84c..  intergenic  DIRS-15_DR         LTR       Ngaro</span></span>
<span><span class="co">## 6 200174837548a59bfeca..  intergenic  DIRS-15_DR         LTR       Ngaro</span></span>
<span><span class="co">##           N                    idx                     seq</span></span>
<span><span class="co">##   &lt;integer&gt;            &lt;character&gt;          &lt;DNAStringSet&gt;</span></span>
<span><span class="co">## 1         1 chr25:14255135-14255.. GTCAGTGATC...TTCCCACCAG</span></span>
<span><span class="co">## 2         1 chr25:12588233-12588.. GAGAGAAGCA...ACCAGGCGGG</span></span>
<span><span class="co">## 3         1 chr25:13719473-13719.. CTTGGTTATG...CCAGTCACTG</span></span>
<span><span class="co">## 4         1 chr25:14323030-14323.. TACGGACAAT...TACCTACTAC</span></span>
<span><span class="co">## 5         1 chr25:14484863-14484.. CAGCGTTGGG...CCCAACACTG</span></span>
<span><span class="co">## 6         1 chr25:14867720-14867.. GGGCGTCACT...TCGCACCCTG</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="section-sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#section-sessioninfo"></a>
</h2>
<div class="sourceCode" id="section-cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2023-11-22 r85609)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] pheatmap_1.0.12                        </span></span>
<span><span class="co">##  [2] EnhancedVolcano_1.21.0                 </span></span>
<span><span class="co">##  [3] ggrepel_0.9.4                          </span></span>
<span><span class="co">##  [4] ggplot2_3.4.4                          </span></span>
<span><span class="co">##  [5] TxDb.Drerio.UCSC.danRer11.refGene_3.4.6</span></span>
<span><span class="co">##  [6] GenomicFeatures_1.55.1                 </span></span>
<span><span class="co">##  [7] AnnotationDbi_1.65.2                   </span></span>
<span><span class="co">##  [8] Biobase_2.63.0                         </span></span>
<span><span class="co">##  [9] BSgenome.Drerio.UCSC.danRer11_1.4.2    </span></span>
<span><span class="co">## [10] BSgenome_1.71.1                        </span></span>
<span><span class="co">## [11] BiocIO_1.13.0                          </span></span>
<span><span class="co">## [12] Biostrings_2.71.1                      </span></span>
<span><span class="co">## [13] XVector_0.43.0                         </span></span>
<span><span class="co">## [14] AnnotationHub_3.11.0                   </span></span>
<span><span class="co">## [15] BiocFileCache_2.11.1                   </span></span>
<span><span class="co">## [16] dbplyr_2.4.0                           </span></span>
<span><span class="co">## [17] edgeR_4.1.2                            </span></span>
<span><span class="co">## [18] limma_3.59.1                           </span></span>
<span><span class="co">## [19] rtracklayer_1.63.0                     </span></span>
<span><span class="co">## [20] GenomicRanges_1.55.1                   </span></span>
<span><span class="co">## [21] GenomeInfoDb_1.39.1                    </span></span>
<span><span class="co">## [22] IRanges_2.37.0                         </span></span>
<span><span class="co">## [23] S4Vectors_0.41.2                       </span></span>
<span><span class="co">## [24] BiocGenerics_0.49.1                    </span></span>
<span><span class="co">## [25] ratte_0.0.3                            </span></span>
<span><span class="co">## [26] BiocStyle_2.31.0                       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3            jsonlite_1.8.7               </span></span>
<span><span class="co">##   [3] magrittr_2.0.3                farver_2.1.1                 </span></span>
<span><span class="co">##   [5] rmarkdown_2.25                fs_1.6.3                     </span></span>
<span><span class="co">##   [7] zlibbioc_1.49.0               ragg_1.2.6                   </span></span>
<span><span class="co">##   [9] vctrs_0.6.4                   memoise_2.0.1                </span></span>
<span><span class="co">##  [11] Rsamtools_2.19.2              RCurl_1.98-1.13              </span></span>
<span><span class="co">##  [13] htmltools_0.5.7               S4Arrays_1.3.0               </span></span>
<span><span class="co">##  [15] progress_1.2.2                curl_5.1.0                   </span></span>
<span><span class="co">##  [17] SparseArray_1.3.1             sass_0.4.7                   </span></span>
<span><span class="co">##  [19] bslib_0.6.0                   desc_1.4.2                   </span></span>
<span><span class="co">##  [21] cachem_1.0.8                  GenomicAlignments_1.39.0     </span></span>
<span><span class="co">##  [23] mime_0.12                     lifecycle_1.0.4              </span></span>
<span><span class="co">##  [25] pkgconfig_2.0.3               Matrix_1.6-3                 </span></span>
<span><span class="co">##  [27] R6_2.5.1                      fastmap_1.1.1                </span></span>
<span><span class="co">##  [29] GenomeInfoDbData_1.2.11       MatrixGenerics_1.15.0        </span></span>
<span><span class="co">##  [31] shiny_1.8.0                   digest_0.6.33                </span></span>
<span><span class="co">##  [33] colorspace_2.1-0              rprojroot_2.0.4              </span></span>
<span><span class="co">##  [35] textshaping_0.3.7             RSQLite_2.3.3                </span></span>
<span><span class="co">##  [37] labeling_0.4.3                filelock_1.0.2               </span></span>
<span><span class="co">##  [39] fansi_1.0.5                   httr_1.4.7                   </span></span>
<span><span class="co">##  [41] abind_1.4-5                   compiler_4.4.0               </span></span>
<span><span class="co">##  [43] bit64_4.0.5                   withr_2.5.2                  </span></span>
<span><span class="co">##  [45] BiocParallel_1.37.0           DBI_1.1.3                    </span></span>
<span><span class="co">##  [47] highr_0.10                    biomaRt_2.59.0               </span></span>
<span><span class="co">##  [49] rappdirs_0.3.3                DelayedArray_0.29.0          </span></span>
<span><span class="co">##  [51] rjson_0.2.21                  tools_4.4.0                  </span></span>
<span><span class="co">##  [53] interactiveDisplayBase_1.41.0 httpuv_1.6.12                </span></span>
<span><span class="co">##  [55] glue_1.6.2                    restfulr_0.0.15              </span></span>
<span><span class="co">##  [57] promises_1.2.1                grid_4.4.0                   </span></span>
<span><span class="co">##  [59] generics_0.1.3                gtable_0.3.4                 </span></span>
<span><span class="co">##  [61] tzdb_0.4.0                    data.table_1.14.8            </span></span>
<span><span class="co">##  [63] hms_1.1.3                     xml2_1.3.5                   </span></span>
<span><span class="co">##  [65] utf8_1.2.4                    BiocVersion_3.19.1           </span></span>
<span><span class="co">##  [67] pillar_1.9.0                  stringr_1.5.1                </span></span>
<span><span class="co">##  [69] later_1.3.1                   splines_4.4.0                </span></span>
<span><span class="co">##  [71] dplyr_1.1.4                   lattice_0.22-5               </span></span>
<span><span class="co">##  [73] bit_4.0.5                     tidyselect_1.2.0             </span></span>
<span><span class="co">##  [75] locfit_1.5-9.8                knitr_1.45                   </span></span>
<span><span class="co">##  [77] bookdown_0.36                 SummarizedExperiment_1.33.0  </span></span>
<span><span class="co">##  [79] xfun_0.41                     statmod_1.5.0                </span></span>
<span><span class="co">##  [81] matrixStats_1.1.0             stringi_1.8.2                </span></span>
<span><span class="co">##  [83] yaml_2.3.7                    evaluate_0.23                </span></span>
<span><span class="co">##  [85] codetools_0.2-19              tibble_3.2.1                 </span></span>
<span><span class="co">##  [87] BiocManager_1.30.22           cli_3.6.1                    </span></span>
<span><span class="co">##  [89] xtable_1.8-4                  systemfonts_1.0.5            </span></span>
<span><span class="co">##  [91] munsell_0.5.0                 jquerylib_0.1.4              </span></span>
<span><span class="co">##  [93] Rcpp_1.0.11                   png_0.1-8                    </span></span>
<span><span class="co">##  [95] XML_3.99-0.15                 parallel_4.4.0               </span></span>
<span><span class="co">##  [97] ellipsis_0.3.2                pkgdown_2.0.7                </span></span>
<span><span class="co">##  [99] readr_2.1.4                   blob_1.2.4                   </span></span>
<span><span class="co">## [101] prettyunits_1.2.0             bitops_1.0-7                 </span></span>
<span><span class="co">## [103] scales_1.2.1                  purrr_1.0.2                  </span></span>
<span><span class="co">## [105] crayon_1.5.2                  rlang_1.1.2                  </span></span>
<span><span class="co">## [107] KEGGREST_1.43.0</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="section-references">References<a class="anchor" aria-label="anchor" href="#section-references"></a>
</h2>
<div id="section-refs" class="references csl-bib-body" line-spacing="2">
<div id="section-ref-kong2019transposable" class="csl-entry">
<div class="csl-left-margin">1. </div>
<div class="csl-right-inline">Kong, Y. <em>et al.</em> Transposable element
expression in tumors is associated with immune infiltration and
increased antigenicity. <em>Nature communications</em>
<strong>10,</strong> 5228 (2019).</div>
</div>
<div id="section-ref-thompson2020identification" class="csl-entry">
<div class="csl-left-margin">2. </div>
<div class="csl-right-inline">Thompson, J. D. <em>et al.</em> Identification
and requirements of enhancers that direct gene expression during
zebrafish fin regeneration. <em>Development</em> <strong>147,</strong>
dev191262 (2020).</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
